### PR TITLE
feat: tooling for L2 and transfer tools deployment

### PR DESCRIPTION
**Pull request by pcarranzav**

### CODE REVIEW

**Comment by tmigone** [comment]
looks good but can we update the readme? The operations section changes significantly with this PR (and if you want you can add an entry to the contracts section for the transfer tool)

**Review by tmigone — APPROVED** [review]

File: deploy/4_l1_transfer_tool.ts [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+
+const func: DeployFunction = async function (hre: HardhatRuntimeEnvironment) {
+  const { deployer } = await hre.getNamedAccounts()
+
+  // Get the addresses we will use
+  const tokenAddress = await promptContractAddress('L1 GRT', logger)
+  if (!tokenAddress) {
+    logger.warn('No token address provided')
+    process.exit(1)
+  }
```
Comment (tmigone) [inline deploy/4_l1_transfer_tool.ts:28]: nit but the prompt and the check for non null address could be wrapped into a function to tidy up a bit.

File: deploy/4_l1_transfer_tool.ts [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+  // the OpenZeppelin hardhat-upgrades tooling, and save the deployment manually.
+  
+  // TODO modify this to use upgradeProxy if a deployment already exists?
+  logger.info('Deploying L1GraphTokenLockTransferTool proxy...')
+  const transferToolFactory = await ethers.getContractFactory('L1GraphTokenLockTransferTool')
+  const transferTool = (await upgrades.deployProxy(transferToolFactory, [owner], {
+    kind: 'transparent',
+    unsafeAllow: ['state-variable-immutable', 'constructor'],
+    constructorArgs: [tokenAddress, l2Implementation, l1Gateway, l1Staking],
+  })) as L1GraphTokenLockTransferTool
```
Comment (tmigone) [inline deploy/4_l1_transfer_tool.ts:66]: 🧐

**Review by tmigone — APPROVED** [review]

File: deploy/4_l1_transfer_tool.ts (hunk 2 above)
Comment (pcarranzav) [reply deploy/4_l1_transfer_tool.ts:66]: Haha I know this is not pretty, it's a bit hard to combine our existing hardhat-deploy setup with OZ upgrades, and just using hardhat-deploy wouldn't get along well with our use of immutables on the implementations, afaict

File: deploy/4_l1_transfer_tool.ts (hunk 1 above)
Comment (pcarranzav) [reply deploy/4_l1_transfer_tool.ts:28]: hmm yeah it might be a bit cleaner but I'd rather leave it like this for now

**Pull request merged**